modules/nlu/src/backend/language-provider.ts

```diff
@@ -110,20 +208,83 @@ export class RemoteLanguageProvider implements LanguageProvider {
     throw new Error(`No provider could successfully fullfil request "${path}" for lang "${lang}"`)
   }
 
-  async vectorize(tokens: string[], lang: string): Promise<number[][]> {
+  /**
+   * Generates words that don't exist in the vocabulary, but that are built from ngrams of existing vocabulary
+   * @param subsetVocab The tokens to which you want similar tokens to
+   */
+  async generateSimilarJunkWords(subsetVocab: string[]): Promise<string[]> {
+    // from totalVocab compute the cachedKey the closest to what we have
+    // if 75% of the vocabulary is the same, we keep the cache we have instead of rebuilding one
+    const gramset = vocabNGram(subsetVocab)
+    let result: string[] | undefined
+
+    this._junkwordsCache.forEach((junk, vocab) => {
+      if (!result) {
+        const sim = setSimilarity(vocab, gramset)
+        if (sim >= 0.75) {
+          result = junk
+        }
+      }
+    })
+
+    if (!result) {
+      // didn't find any close gramset, let's create a new one
+      const realWords = _.uniq(subsetVocab)
+      const meanWordSize = _.meanBy(realWords, w => w.length)
+      const minJunkSize = Math.max(JUNK_TOKEN_MIN, meanWordSize / 3)
+      const maxJunkSize = Math.min(JUNK_TOKEN_MAX, meanWordSize * 1.5)
+
+      result = _.range(0, JUNK_VOCAB_SIZE).map(() =>
+        _.sampleSize(gramset, _.random(minJunkSize, maxJunkSize, false)).join('')
+      ) // randomly generated words
+
+      this._junkwordsCache.set(gramset, result)
+      this.onJunkWordsCacheChanged()
+    }
+
+    return result
+  }
+
+  async vectorize(tokens: string[], lang: string): Promise<Float32Array[]> {
     if (!tokens.length) {
       return []
     }
 
-    const text = tokens.reduce((a, b) => a + b, '')
-    const cacheKey = `${lang}_${encodeURI(text)}`
+    const vectors: Float32Array[] = Array(tokens.length)
+    const idxToFetch: number[] = [] // the tokens we need to fetch remotely
+    const getCacheKey = (t: string) => `${lang}_${encodeURI(t)}`
```

Comment:
This is not, encoreURIComponent was an autocomplete type.
Text is provided by the user, it might include any char, I thought it would be safer to use encodeURI so that we don't use spaces, etc in the cache key.